### PR TITLE
Start emulator from checkpoint

### PR DESCRIPTION
**Pull request by janezpodhostnik**
## Description

Start the emulator on top of a checkpoint.

______

For contributor use:

- [ ] Targeted PR against `master` branch
- [ ] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [ ] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

**Review by turbolent — COMMENTED** [review]
Nice! This will be super useful 👍

**Review by bluesign — APPROVED** [review]

File: storage/checkpoint/checkpoint.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/onflow/flow-go/module/metrics"
+
+	"github.com/onflow/flow-emulator/storage"
+	"github.com/onflow/flow-emulator/storage/memstore"
+)
+
+// Store is jus a memstore, but the starting state is loaded from a checkpoint folder
+// any new blocks exist in memory only and are not persisted to disk.
+type Store struct {
+	*memstore.Store
```
Comment (sideninja) [inline storage/checkpoint/checkpoint.go:45]: Could this work with non-mem-store?

File: server/server.go [hunk 2]
```diff
@@ -332,6 +339,20 @@ func configureStorage(logger *zerolog.Logger, conf *Config) (storageProvider sto
 		}
 	}
 
+	if conf.CheckpointPath != "" || conf.StateHash != "" {
```
Comment (sideninja) [inline server/server.go:342]: If this requires mem-store it should provide an error if used with persist flag

**Review by sideninja — APPROVED** [review]
Awesome feature!

File: cmd/emulator/start/start.go [hunk 3]
```diff
@@ -76,6 +76,8 @@ type Config struct {
 	LegacyContractUpgradeEnabled bool          `default:"false" flag:"legacy-upgrade" info:"enable Cadence legacy contract upgrade"`
 	StartBlockHeight             uint64        `default:"0" flag:"start-block-height" info:"block height to start the emulator at. only valid when forking Mainnet or Testnet"`
 	RPCHost                      string        `default:"" flag:"rpc-host" info:"rpc host to query when forking Mainnet or Testnet"`
+	CheckpointPath               string        `default:"" flag:"checkpoint-dir" info:"checkpoint directory to load the emulator state from"`
```
Comment (sideninja) [inline cmd/emulator/start/start.go:79]: Update readme with this flags 🙏

File: storage/checkpoint/checkpoint.go (hunk 1 above)
Comment (janezpodhostnik) [reply storage/checkpoint/checkpoint.go:45]: theoretically yes, but I saw no use case for it.

**Review by turbolent — APPROVED** [review]

**Pull request merged**